README.md
[![Explorer Demo](./documents/explorer-demo.gif)](https://www.youtube.com/watch?v=6cCjX24hKwo)

This project is a web porting for [NVlabs' StyleGAN2](https://github.com/NVlabs/stylegan2), to facilitate exploring all kinds characteristic of StyleGAN networks.

Thanks for [NVlabs](https://github.com/NVlabs)' excellent work.

## Features

### Explorer

See how the result image response to changes of latent code & psi.

![Projector Demo 1](./documents/explorer-illustration.png)

### Projector

Test the projection from image to latent code. Left is target image, right is result from generator model. Total step count and yielding interval can be configured in this page. And another hyperparameter `regularize noise weight` can be configured statically, see [Environment Configurations](#environment-configurations).

You can save the projection result in a zip package, and this page can accept projector zip file dropping, so this feature enable you to share your projector result to others.

![Projector Demo 1](./documents/projector-demo1.png)
![Projector Demo 2](./documents/projector-demo2.png)

#### W latent space extension

We added an env switch of `UNIFORM_LATENTS` to denote using *uniform* or *separated* W latent code when projecting image. This is the results comparison (center is the target):

<p align="center">
<img src="./documents/ffhq-cover-girl-projection-w.jpg" alt="W latents result" valign="middle" /> &nbsp;&nbsp;&#x21e8;&nbsp;&nbsp; 
<img src="./documents/ffhq-cover-girl.jpg" alt="target image" valign="middle" /> &nbsp;&nbsp;&#x21e6;&nbsp;&nbsp; 
<img src="./documents/ffhq-cover-girl-projection-w+.jpg" alt="W+ latents result" valign="middle" />
</p>

#### Projection animation exporting

You can also export image projection result sequence as a gif animation:

<p align="center">
<img src="./documents/cat-roundeyes.png" valign="middle" /> &nbsp;&nbsp;&#x21e6;&nbsp;&nbsp; <img src="./documents/cat-roundeyes-projection-1000.gif" valign="middle" />
</p>

#### Face image pose alignment

Dataset of [FFHQ](https://github.com/NVlabs/ffhq-dataset)'s generation has a crop process to align face area.<sup>[see paper](https://arxiv.org/abs/1710.10196), appendix C.</sup> So the output distribution of StyleGAN model learned on FFHQ has a strong prior tendency on features position. We observed that many face images projection suffers semantic mistakes, e.g. erasing original eyes and transforming eyebrow into eyes during projection fitting (however you can get a similar face at last, but it may yield freak results when you manipulate the latent code). Finally we figured out that mainly caused by unalignment with training dataset prior distribution.

Then we import the [face-api](https://github.com/justadudewhohacks/face-api.js) to measure and align human face images as below:

<p align="center" style="white-space: nowrap">
	<img src="./documents/face-crop-origin.jpg" valign="middle" width="240" />
	- &#x1F642; &#x2192;
	<img src="./documents/face-mark-demo.png" valign="middle" width="240" />
	- &#x2704; &#x2192;
	<img src="./documents/face-crop.jpg" valign="middle" width="240" />
</p>

*Gratefulness for the authorization by [@芈砾](https://www.douban.com/people/43619507/) to use his nice [opus](https://www.douban.com/photos/photo/2575639110/).*

Click the button [&#x1F642;] after target image loaded, if the face detection succeed, you will get the face landmark and proposed crop box. The detection result may be not very accurate, now you can adjust 3 anchor marks manually to align left eye (red), right eye (green) and mouth (blue). Then click button [&#x2704;] to apply the crop.

### Merger

Once you get some latent codes by projector or turning, you can test to mix features by interpolating latent values on every W layer. This is a demo.

![Merger Demo](./documents/merger-demo.gif)

The pair of top-left images are the source to merge, press **Ctrl+V** in the hash box below either image to paste input latent code via clipboard, and **Ctrl+C** on the right blank area to copy result latent code.

### Mapping Network Research

[![mapping plot](/documents/mapping-plot.png)](https://k-l-lambda.github.io/2020/02/10/stylegan-mapping/)

I attempt to explore the StyleGAN mapping network high-dimensional terrain aspect, [**read this article**](https://k-l-lambda.github.io/2020/02/10/stylegan-mapping/) for details.

## Usage

Run the web server:

```.bash
python ./http_server.py
```

If this works, open *http://localhost:8186* in your browser.

To ensure it working, please read the following requirements before do this.

## Requirements

### Python

Install requirement libraries with pip, reference to [requirements.txt](./requirements.txt).

### Network Files

Before run the web server, StyleGAN2 pre-trained network files must be placed in local disk (recommended the folder `models`). You can download network files following to [StyleGAN2's code](https://github.com/NVlabs/stylegan2/blob/master/pretrained_networks.py).

For memory reason, only one generator model can be loaded when running the web server. Network file paths can be configured by env variables. Create a file named `.env.local` under project root to configure chosen model and network file paths. Network file name/paths are configured in key-value style, e.g.:

<a name="model-paths-example"></a>
```.env
MODEL_NAME=ffhq		# ffhq is the default value, so this line can be ignored 

MODEL_PATH_ffhq=./models/stylegan2-ffhq-config-f.pkl
MODEL_PATH_cat=./models/stylegan2-cat-config-f.pkl
# And so on...
```

Alternately, you can also choose generator model name by start command argument, e.g.:

```.bash
python ./http_server.py cat
```

Or, for nodejs developer:

```.bash
yarn start cat
```

Besides generators, the network *LPIPS* is required when run image projector, the default local path is `./models/vgg16_zhang_perceptual.pkl`, [download link](https://drive.google.com/uc?id=1N2-m9qszOeVC9Tq77WxsLnuWwOedQiD2). You can also change local path by env variable `MODEL_PATH_LPIPS`.

### For Windows

According to [StyleGAN2 README.md](https://github.com/NVlabs/stylegan2#requirements), here are our additional help instructions:

* MSVC

	**NOTE**: Visual Studio 2019 Community Edition seems not compatible with CUDA 10.0, Visual Studio 2017 works.

	Append the actual msvc binary directory (find in your own disk) into `dnnlib/tflib/custom_ops.py`, the array of `compiler_bindir_search_path`. For example:

	```patch
	-	'C:/Program Files (x86)/Microsoft Visual Studio 14.0/vc/bin',
	+	'C:/Program Files (x86)/Microsoft Visual Studio/2017/BuildTools/VC/Tools/MSVC/14.16.27023/bin/Hostx64/x64',
	```

* NVCC

	To test if nvcc is configured properly, dowload [test_nvcc.cu](https://github.com/NVlabs/stylegan2/blob/master/test_nvcc.cu) in StyleGAN2 project. And the test command should specify binary path:

	```.bash
	nvcc test_nvcc.cu -o test_nvcc -run -ccbin "C:\Program Files (x86)\Microsoft VisualStudio\2017\BuildTools\VC\Tools\MSVC\14.16.27023\bin\Hostx64\x64"
	```

	Actual path to different msvc edition may have difference in detail. If this succeed, it will build a file named `test_nvcc.exe`.

* Tips for tensorflow 1.15

	Tensorflow 1.15 can work under Windows, but *NVCC* compiling may encounter C++ including path problem. Here is an easy workaround: make a symbolic link in python installation directory `Python36\Lib\site-packages\tensorflow_core`:

	```.bash
	mklink /J tensorflow tensorflow_core
	```

* Tips for tensorflow 2.x

	**Tensorflow 2.0+ can work now!** I have solved the compatibility issues with TF2 already, including some modification of code bundled in pickle.
	Except one problem on Windows, if you encountered this:

	>	C:/Users/xxx/AppData/Local/Programs/Python/Python36/lib/site-packages/tensorflow/include\unsupported/Eigen/CXX11/Tensor(74): fatal error C1083: Cannot open include file: 'unistd.h': No such file or directory

	Just open this file and comment out this line simply:

	> #include <unistd.h>

	It seems a bug of tensorflow, and I have committed an [issue](https://github.com/tensorflow/tensorflow/issues/39829#event-3369800641) for them.

* cudafe++ issue

	If you encountered python console error like:
	
	```
	nvcc error : 'cudafe++' died with status 0xC0000005 (ACCESS_VIOLATION)
	```

	That may caused by a bug from CUDA 10.0, you can fix this issue by replacing your `cudafe++.exe` file in CUDA program bin directory by the same name file from CUDA 10.1 or higher version. And you are welcome to download my [backup](./documents/cudafe++.zip) to avoid install a whole new version CUDA.

## Environment Configurations

To manage environment variables conveniently, create a configuration file named `.env.local`. All avaiable env list:

Key							| Description							| Default Value
:--							| :--									| :--
**HTTP_HOST**				| Web server host.						| *127.0.0.1*
**HTTP_PORT**				| Web server port.						| *8186*
**MODEL_NAME**				| Name for the generator model to load, this can be overwrite by the first argument of *[start script](./package.json#L7)*.	| *ffhq*
**MODEL_PATH_LPIPS**		| File path for LPIPS model.			| *./models/vgg16_zhang_perceptual.pkl*
**MODEL_PATH_***			| Generator network file path dictionary. See [examples](#model-paths-example).	|
**REGULARIZE_NOISE_WEIGHT**	| Projector training hyperparameter. Float.	| *1e5*
**INITIAL_NOISE_FACTOR**	| Projector training hyperparameter. Float.	| *0.05*
**EUCLIDEAN_DIST_WEIGHT**	| Projector training hyperparameter. Float.	| *1*
**REGULARIZE_MAGNITUDE_WEIGHT**	| Projector training hyperparameter. Float.	| *0*
**UNIFORM_LATENTS**			| Use uniform latents for all feature layers (consistent with origin StyleGAN2 paper). Boolean, 0 or 1	| *0*


changelog.txt

TODO:
	rotate vector by source-target input
	project from a specified code



2020.5.24	K.L.	<k.l.lambda@gmail.com>

	+	Replace tf importing by tensoflow.compat.v1.


2020.4.25	K.L.	<k.l.lambda@gmail.com>

	+	mapppingViewer.vue: downloadSourceDataTSV added.


2020.2.8	K.L.	<k.l.lambda@gmail.com>

	+	mapppingViewer.vue: compressed images downloading added.


2020.2.6	K.L.	<k.l.lambda@gmail.com>

	+	mapppingViewer.vue: source data & result images download methods implemented.


2020.1.25	K.L.	<k.l.lambda@gmail.com>

	+	circlePlot.vue: plot circle sliced 3 dimensions.


2020.1.23	K.L.	<k.l.lambda@gmail.com>

	+	mapppingViewer.vue: mapping from Z to W implemented.

	+	mapppingViewer.vue: circleSamplePoints added.

	+	vectorInput.vue: latent code copy added.

	+	vectorInput.vue created.
	+	circlePlot.vue created.

	+	Entry of mappingViewer created.


2020.1.22	K.L.	<k.l.lambda@gmail.com>

	+	projector.vue: edit face mark point without detection.

	+	projector.vue: added face detection remove button.


2020.1.21	K.L.	<k.l.lambda@gmail.com>

	+	Flick outline when copying latent codes.


2020.1.19	K.L.	<k.l.lambda@gmail.com>

	>	projector.vue: fixed projected sequence number when skip steps.


2020.1.18	K.L.	<k.l.lambda@gmail.com>

	+	projector.vue: target image download link added.

	+	projector.vue: move face ref marks manually.

	+	projector.vue: mirror edge neighbor images in face crop added.

	+	projector.vue: cropFace implemented.

	+	projector.vue: face crop graph added.

	+	projector.vue: face landmark detection added.

	+	projector.vue: face-api weights loading added.

	+	projector.vue: imported face-api.

	+	http_server.py: separated static folder from dist.


2020.1.17	K.L.	<k.l.lambda@gmail.com>

	+	index.vue: added button of mapping Z to W.

	+	g-view.vue: package file drag & drop added.

	+	projector.py: added latent magnitude regularization loss item.

	+	projector.vue: added index number on animation.

	+	merger.vue: added result download button.


2020.1.16	K.L.	<k.l.lambda@gmail.com>

	+	merger.vue: download result image method added.

	>	projector.vue: refined yield image sequence click behavior.

	>	projector.vue: fixed field type of projectSteps & projectYieldInterval.


2020.1.12	K.L.	<k.l.lambda@gmail.com>

	+	projector.vue: batch projection implemented.

	+	index.vue: show W latent magnitude.

	+	index.vue: added latent URL copy shortcut.

	+	g-view.vue: accept Z latent code input.

	+	http_server.py: added argument of psi for mapZtoW.

	+	http_server.py: mapZtoW added.

	+	merger.vue: source swapping button added.

	+	merger.vue: added 3 formula options.


2020.1.11	K.L.	<k.l.lambda@gmail.com>

	-	index.vue: removed lerp button for W latents.

	+	merger.vue: cache result latent code to reduce generation requests.

	+	navigator.vue: added link for merger.

	+	merger.vue: store source codes in session storage.

	+	g-view.vue: result image added.

	+	g-view.vue: latent copy added.

	+	g-view.vue created.

	+	index.vue: use latents URL for clipboard copy & paste.

	+	Entry of merger added.


2020.1.9	K.L.	<k.l.lambda@gmail.com>

	*!	Eulidean loss lead to blur result.

	+	projector.py: use origin generator resolution for eulidean distance loss.


2020.1.7	K.L.	<k.l.lambda@gmail.com>

	+	projector.py: added eulidean distance on loss function.


2020.1.6	K.L.	<k.l.lambda@gmail.com>

	+	projector.vue: target name control added.


2020.1.4	K.L.	<k.l.lambda@gmail.com>

	+	projector.vue: animation panel implemented.

	+	projector.vue: save all latent codes in archive package.


2020.1.3	K.L.	<k.l.lambda@gmail.com>

	>	index.vue: fixed latent layer dimensions issue.

	>	projector.vue: fixed latent layer dimensions issue.

	+	projector.vue: makeAnimation added.


2020.1.2	K.L.	<k.l.lambda@gmail.com>

	+	index.vue: added option of extend feature.

	+	projector.vue: use fix16 extended latent code.

	+	projector.py: added parameter of uniform_latents.


2019.12.29	K.L.	<k.l.lambda@gmail.com>

	>	app: added some assistive elements.

	+	app: added navitator on pages.


2019.12.27	K.L.	<k.l.lambda@gmail.com>

	+	index.vue: latent code copy & paste added.

	+	http_server.py: added env variable of REGULARIZE_NOISE_WEIGHT.


2019.12.26	K.L.	<k.l.lambda@gmail.com>

	>	projector.vue: adjusted movement threshold.

	+	index.vue: lerp latents in W space implemented.

	+	index.vue: refined latent type control.


2019.12.25	K.L.	<k.l.lambda@gmail.com>

	>	projector.vue: refined layouts.

	+	projector.vue: result package loading implemented.

	+	projector.vue: result package saving implemented.

	+	projector.vue: added option to hide trivial sequence items.

	>	index.vue: refined feature bars.

	>	index.vue: refined controls.

	+	http_server.py: W synthesis added.

	*!	Yarn run on a symbolic link folder may result in path issue.


2019.12.24	K.L.	<k.l.lambda@gmail.com>

	+	favicon.ico created.

	+	http_server.py: resize target image in project handler.

	+	latentCode.js created.

	+	projector.vue: image drop added.

	+	projector.vue: generator links added.

	>	Renamed file: http-server.py -> http_server.py.

	+	http-server.py: output step in project handler.

	+	http-server.py: project handler implemented.


2019.12.23	K.L.	<k.l.lambda@gmail.com>

	+	projector.vue: generator fetching of projectImage added.

	+	app: entry of projector created.

	+	http-server.py: handler of project added.

	+	test-projector.py: progressively projecting implemneted.

	+	test-projector.py created.


2019.12.21	K.L.	<k.l.lambda@gmail.com>

	>	http-server.py: use single thread Flask to fix tensorflow variable missing issue.

	*!	Flask multi-thread result in tensorflow variable missing, when model run parameter changed, in dnnlib.Network.get_output_for.

	+	http-server.py: image generation implemented.

	+	http-server.py: StyleGAN2 model loading added.

	+	http-server.py: http server based on Flask created.

	+	test.py: image generation for StyleGAN2 implemented.

	+	Branch of v2 setup.



http_server.py

import os
import sys
import io
import time
from dotenv import load_dotenv
import flask
import pickle
import PIL.Image
import base64
import numpy as np
import tensorflow.compat.v1 as tf
tf.disable_v2_behavior()
from threading import Lock
import json
import re

import dnnlib.tflib
from training import misc
from projector import Projector
import latentCode



load_dotenv(dotenv_path = './.env.local')
load_dotenv()


g_Gs = None
g_Synthesis = None
g_Lpips = None
g_Projector = None
g_Session = None
g_LoadingMutex = Lock()


def loadGs():
	with g_LoadingMutex:
		global g_Gs, g_Synthesis
		if g_Gs:
			return g_Gs, g_Synthesis

		global model_name

		model_path = os.environ.get('MODEL_PATH_%s' % model_name)

		if model_path is None:
			print('invalid model name:', model_path)
			return

		global g_Session
		if g_Session is None:
			print('Initializing dnnlib...')
			dnnlib.tflib.init_tf()
			g_Session = tf.get_default_session()

		print('Loading model %s ...' % model_name)

		with open(model_path, 'rb') as f:
			with g_Session.as_default():
				Gi, Di, Gs = pickle.load(f)
				g_Gs = Gs

				## burn one
				#images = Gs.run(np.zeros((1, Gs.input_shape[1])), None, truncation_psi = 0.7, output_transform = dict(func = dnnlib.tflib.convert_images_to_uint8, nchw_to_nhwc = True))
				#print('images:', g_Session, images)

				#print('Gs.components.synthesis.input_shape:', Gs.components.synthesis.input_shape)
				global g_dLatentsIn
				g_dLatentsIn = tf.placeholder(tf.float32, [Gs.components.synthesis.input_shape[1] * Gs.input_shape[1]])
				dlatents_expr = tf.reshape(g_dLatentsIn, [1, Gs.components.synthesis.input_shape[1], Gs.input_shape[1]])
				g_Synthesis = Gs.components.synthesis.get_output_for(dlatents_expr, randomize_noise = False)

	return g_Gs, g_Synthesis


def loadLpips():
	with g_LoadingMutex:
		global g_Lpips
		if g_Lpips:
			return g_Lpips

		model_path = os.environ.get('MODEL_PATH_LPIPS')

		if model_path is None:
			print('invalid model name:', model_path)
			return

		global g_Session
		if g_Session is None:
			print('Initializing dnnlib...')
			dnnlib.tflib.init_tf()
			g_Session = tf.get_default_session()

		print('Loading model lpips ...')

		with open(model_path, 'rb') as f:
			with g_Session.as_default():
				lpips = pickle.load(f)
				g_Lpips = lpips

	return g_Lpips


def loadProjector():
	global g_Projector
	if g_Projector:
		return g_Projector

	gs, _ = loadGs()
	lpips = loadLpips()

	g_Projector = Projector()
	g_Projector.regularize_noise_weight = float(os.environ.get('REGULARIZE_NOISE_WEIGHT', 1e5))
	g_Projector.initial_noise_factor = float(os.environ.get('INITIAL_NOISE_FACTOR', 0.05))
	g_Projector.uniform_latents = int(os.environ.get('UNIFORM_LATENTS', 0)) > 0
	g_Projector.euclidean_dist_weight = float(os.environ.get('EUCLIDEAN_DIST_WEIGHT', 1))
	g_Projector.regularize_magnitude_weight = float(os.environ.get('REGULARIZE_MAGNITUDE_WEIGHT', 0))
	g_Projector.set_network(gs, lpips)

	return g_Projector


app = flask.Flask(__name__, static_url_path = '', static_folder = './static')


DIST_DIR = './dist'


@app.route('/bundles/<path:filename>')
def bundle(filename):
	if re.match(r'.*\.bundle\.js$', filename):
		return flask.send_from_directory(DIST_DIR, filename)

	flask.abort(404, 'Invalid request path.')


pageRouters = {
	'/':				'index.html',
	'/projector/':		'projector.html',
	'/merger/':			'merger.html',
	'/mapping-viewer/':	'mappingViewer.html',
}
for path in pageRouters:
	def getHandler(filename):
		return lambda: flask.send_from_directory(DIST_DIR, filename)

	app.route(path, endpoint = 'handler' + path)(getHandler(pageRouters[path]))


@app.route('/spec', methods=['GET'])
def spec():
	global model_name
	model, _ = loadGs()

	return dict(
		model = model_name,
		latents_dimensions = model.input_shape[1],
		image_shape = model.output_shape,
		synthesis_input_shape = model.components.synthesis.input_shape)


@app.route('/map-z-w', methods=['GET'])
def mapZtoW():
	zStr = flask.request.args.get('z')
	psi = flask.request.args.get('psi')
	psi = psi and float(psi)

	gs, _ = loadGs()
	latent_len = gs.input_shape[1]

	z = latentCode.decodeFloat32(zStr, latent_len).reshape([1, latent_len])
	w = gs.components.mapping.run(z, None)[:, :1, :].reshape([latent_len])

	if psi is not None:
		proj = loadProjector()
		avg = proj._dlatent_avg.reshape([latent_len])

		w = avg + psi * (w - avg)

	return latentCode.encodeFloat32(w)


@app.route('/generate', methods=['GET'])
def generate():
	latentsStr = flask.request.args.get('latents')
	latentsStrX = flask.request.args.get('xlatents')
	psi = float(flask.request.args.get('psi', 0.5))
	#use_noise = bool(flask.request.args.get('use_noise', True))
	randomize_noise = int(flask.request.args.get('randomize_noise', 0))
	fromW = int(flask.request.args.get('fromW', 0))

	global g_Session
	global g_dLatentsIn
	#print('g_Session.1:', g_Session)

	gs, synthesis = loadGs()

	latent_len = gs.input_shape[1]
	if latentsStrX:
		latents = latentCode.decodeFixed16(latentsStrX, g_dLatentsIn.shape[0])
	else:
		latents = latentCode.decodeFloat32(latentsStr, latent_len)

	t0 = time.time()

	# Generate image.
	fmt = dict(func = dnnlib.tflib.convert_images_to_uint8, nchw_to_nhwc = True)
	with g_Session.as_default():
		if fromW != 0:
			#print('latentsStr:', latentsStr)
			#print('shapes:', g_dLatentsIn.shape, latents.shape)

			if latents.shape[0] < g_dLatentsIn.shape[0]:
				latents = np.tile(latents, g_dLatentsIn.shape[0] // latents.shape[0])
			images = dnnlib.tflib.run(synthesis, {g_dLatentsIn: latents})
			image = misc.convert_to_pil_image(misc.create_image_grid(images), drange = [-1,1])
		else:
			latents = latents.reshape([1, latent_len])
			images = gs.run(latents, None, truncation_psi = psi, randomize_noise = randomize_noise != 0, output_transform = fmt)
			image = PIL.Image.fromarray(images[0], 'RGB')

	print('generation cost:', time.time() - t0)

	# encode to PNG
	fp = io.BytesIO()
	image.save(fp, PIL.Image.registered_extensions()['.png'])

	return flask.Response(fp.getvalue(), mimetype = 'image/png')


#LPIPS_IMAGE_SHAPE = tuple(map(int, os.environ.get('LPIPS_IMAGE_SHAPE', '256,256').split(',')))


@app.route('/project', methods=['POST'])
def project():
	steps = int(flask.request.args.get('steps', 1000))
	yieldInterval = int(flask.request.args.get('yieldInterval', 10))
	#regularizeNoiseWeight = float(flask.request.args.get('regularizeNoiseWeight', 1e5))

	imageFile = flask.request.files.get('image')
	if not imageFile:
		flask.abort(400, 'image field is requested.')

	Gs, _ = loadGs()
	image = PIL.Image.open(imageFile.stream).resize((Gs.output_shape[2], Gs.output_shape[3]), PIL.Image.ANTIALIAS)

	image_array = np.array(image)[:, :, :3].swapaxes(0, 2).swapaxes(1, 2)
	image_array = misc.adjust_dynamic_range(image_array, [0, 255], [-1, 1])

	#print('shape:', image_array.shape)

	def gen():
		proj = loadProjector()
		#proj.regularize_noise_weight = regularizeNoiseWeight
		proj.start([image_array])
		for step in proj.runSteps(steps):
			print('\rProjecting: %d / %d' % (step, steps), end = '', flush = True)

			if step % yieldInterval == 0:
				dlatents = proj.get_dlatents()
				images = proj.get_images()
				pilImage = misc.convert_to_pil_image(misc.create_image_grid(images), drange = [-1,1])

				fp = io.BytesIO()
				pilImage.save(fp, PIL.Image.registered_extensions()['.png'])

				imgUrl = 'data:image/png;base64,%s' % base64.b64encode(fp.getvalue()).decode('ascii')

				#latentsList = list(dlatents.reshape((-1, dlatents.shape[2])))
				#latentCodes = list(map(lambda latents: latentCode.encodeFloat32(latents).decode('ascii'), latentsList))
				latentCodes = latentCode.encodeFixed16(dlatents.flatten()).decode('ascii')

				yield json.dumps(dict(step = step, img = imgUrl, latentCodes = latentCodes)) + '\n\n'

		print('\rProjecting finished.%s' % (' ' * 8))

	return flask.Response(gen(), mimetype = 'text/plain')


def main(argv):
	global model_name
	model_name = argv[1] if len(argv) > 1 else os.environ.get('MODEL_NAME')

	try:
		app.run(port = int(os.getenv('HTTP_PORT')), host = os.getenv('HTTP_HOST'), threaded = False)
	except:
		print('server interrupted:', sys.exc_info())



if __name__ == "__main__":
	main(sys.argv)
